### ISSUE TITLE
AlreadyRegisteredException: A wizard has already been registered for model: Post

### ISSUE THREAD

**Issue by bxm156** [nephila/djangocms-blog#217]
On my instance of django-cms, I added 2 instances of the djangocms-blog app, under different namespaces. Everything seemed to work together, but I later restarted the server to disable CMS caching, and now I get the following error:

AlreadyRegisteredException: A wizard has already been registered for model: Post

(Not sure if this larger DjangoCMS issue, or related to djangocms_blog, but figured the mention of 'Post' means this would be a good place to start)

Not sure what happened, but since I don't need 2 blog instances (I was just testing stuff), I think I'm going to just drop my database tables and start all over. Thought I would post this here just in case this is an actual bug. 

```
Traceback (most recent call last): 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 174, in get_response response = self.process_exception_by_middleware(e, request) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 172, in get_response response = response.render() 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/response.py", line 160, in render self.content = self.rendered_content 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/response.py", line 137, in rendered_content content = template.render(context, self._request) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/backends/django.py", line 95, in render return self.template.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 206, in render return self._render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 197, in _render return self.nodelist.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 992, in render bit = node.render_annotated(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 959, in render_annotated return self.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/loader_tags.py", line 173, in render return compiled_parent._render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 197, in _render return self.nodelist.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 992, in render bit = node.render_annotated(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 959, in render_annotated return self.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/loader_tags.py", line 173, in render return compiled_parent._render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 197, in _render return self.nodelist.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 992, in render bit = node.render_annotated(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 959, in render_annotated return self.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/classytags/core.py", line 138, in render return self.render_tag(context, **kwargs) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/sekizai/templatetags/sekizai_tags.py", line 83, in render_tag rendered_contents = nodelist.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 992, in render bit = node.render_annotated(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/django/template/base.py", line 959, in render_annotated return self.render(context) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/classytags/core.py", line 138, in render return self.render_tag(context, **kwargs) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/templatetags/cms_tags.py", line 687, in render_tag request.toolbar.post_template_populate() 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/toolbar/toolbar.py", line 313, in post_template_populate self._call_toolbar('post_template_populate') 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/toolbar/toolbar.py", line 351, in _call_toolbar result = getattr(toolbar, func_name)() 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/cms_toolbars.py", line 73, in post_template_populate self.add_wizard_button() 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/cms_toolbars.py", line 109, in add_wizard_button list(entry_choices(user, self.page))) == 0 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/wizards/wizard_pool.py", line 19, in entry_choices for entry in wizard_pool.get_entries(): 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/wizards/wizard_pool.py", line 117, in get_entries self._discover() 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/wizards/wizard_pool.py", line 35, in _discover load('cms_wizards') 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/utils/django_load.py", line 61, in load get_module(app, modname, verbose, failfast) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/utils/django_load.py", line 45, in get_module module = import_module(module_name) 
 File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module __import__(name) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/djangocms_blog/cms_wizards.py", line 53, in <module> wizard_pool.register(post_wizard) 
 File "/home/site/.env/site/local/lib/python2.7/site-packages/cms/wizards/wizard_pool.py", line 80, in register model.__name__) AlreadyRegisteredException: A wizard has already been registered for model: Post
```

**Comment by rt-em** [comment]
I tried to upgrade djangocms-blog app from version 0.6.3 to 0.7.0b1. On the production-like local computer - everything works well, but upgrade of djangocms-blog crushed the production and now I have the same issue. Now I can't edit all the cms pages because of AlreadyRegisteredException error.

**Comment by yakky** [comment]
@bxm156 @rt-em which django and django CMS version are you using?

**Comment by yakky** [comment]
AlreadyRegisteredException is triggered by the CMS when multiple wizards with the same  class id (ie: same memory address in python) are registered. Leftover pyc around when upgrading the blog, maybe?

**Comment by bxm156** [comment]
In my case,  I didn't upgrade anything. I just stopped my websever,  turned
off cms caching via settings.py,  then started it back up.

**Comment by yakky** [comment]
@bxm156 did you mean you just removed ``CACHES`` settings from settings.py? Which django / django CMS / djangocms_blog version are you on?

**Comment by rt-em** [comment]
Django CMS 3.2.0
Django 1.8.8
I delete all Blog configs, except one(wich was connected with posts), delete the blog page in cms pages, delete all the Blog plugins from placeholders using pgadmin. Then I restarted uwsgi and nginx, and after that my site started. Then I add blog page again and set blog apphook to this page - blog now is working. But now I can't add plugins to cms placeholders... I keep digging

**Comment by bxm156** [comment]
Django CMS 3.2.1
Django 1.9

The only change in my code was adding

```
CMS_PAGE_CACHE = False
CMS_PLACEHOLDER_CACHE = False
CMS_PLUGIN_CACHE = False
```
to settings.py

**Comment by rt-em** [comment]
I found that can't add blog plugins inside bootstrap elements of djangocms-cascade plugin where I previously delete blog plugins using pgadmin. I totaly cleaned the placeholders using cms, rebuild the page structure, and add the blog plugins again. For this moment everything works well, but I didn't identified the reason of the crash and I didn't identified wich action led to total recovery... But I hope that information will be usefull for someone facing the same problems

**Comment by yakky** [comment]
@rt-em @bxm156 aldryn-apphooks-config 0.2.7 should fix some of these issues
I opened a issue there to track it. It has little to do with the blog, although I'll improve how the blog can mitigate these issues

**Comment by yakky** [comment]
You can test linked PR

**Comment by yakky** [comment]
Sorry comments / links on wrong issue :/

**Comment by yakky** [comment]
@bxm156 @rt-em I've been unable to reproduce the issue, but commit 344064d should mitigate the isue for production websites. In development the error is triggered to help debugging

**Comment by yakky** [comment]
I opened a pull request with a mitigation strategy for this issue

**Comment by yakky** [comment]
@bxm156 @rt-em I released djangocms_blog 0.7.0b2 which includes the mitigation fix

**Comment by bxm156** [comment]
Thanks @yakky !

**Comment by sephii** [comment]
I'm having the same issue: I have two appconfigs for the blog and I'm getting the same exception.

**Comment by yakky** [comment]
@sephii using version 0.7 final?

**Comment by yakky** [comment]
@sephii are you able to reproduce locally / with a small projec / dataset you can share?

**Comment by sephii** [comment]
I have tested against latest develop. I don't have time right now to reproduce it with a small project but I'll gladly share my current findings, hoping they will be of any use.

I created two different pages attached to two different Sites, and set each of them to a different AppConfig. Because of this, the following code (from `djangocms_blog/cms_wizards.py`) goes two times in the loop (because I have two `BlogConfig` objects):

```python
for config in BlogConfig.objects.all().order_by('namespace'):                         
    new_wizard = type(str(slugify(config.app_title)), (PostWizard,), {})
    new_form = type(str('{0}Form').format(slugify(config.app_title)), (PostWizardForm,    ), {    
        'default_appconfig': config.pk
    })  
    post_wizard = new_wizard(
        title=_('New {0}').format(config.object_name),
        weight=200,
        form=new_form,
        model=Post,
        description=_('Create a new {0} in {1}').format(config.object_name, config.app    _title),    
    )
    try:
        wizard_pool.register(post_wizard)
    except AlreadyRegisteredException:  # pragma: no cover
        if settings.DEBUG:
            raise
```

When running `wizard_pool.register`, the following code is executed (from `cms/wizards/wizard_pool.py`):
```python
 def register(self, entry):
    """
    Registers the provided «entry».

    Raises AlreadyRegisteredException if the entry is already registered.
    """
    assert isinstance(entry, Wizard), u"entry must be an instance of Wizard"
    if self.is_registered(entry, passive=True):
        model = entry.get_model()
        raise AlreadyRegisteredException(
            _(u"A wizard has already been registered for model: %s") %
            model.__name__)
```

The `is_registered` method checks if the `id` attribute of the entry is already present in the registered entries. It will run fine for the 1st wizard but will raise an exception for the second wizard because the `id` property is computed from the `self.__module__` and `self.__class__.__name__` attributes, which are the same for the two wizards created previously.

I hope that's clear and that my analysis is correct. I have absolutely no idea how to solve it though. ;) Maybe by never raising an exception if a wizard is already registered and just ignore it?

**Comment by sephii** [comment]
I found a workaround that avoids patching DjangoCMS itself but I'm really not sure it's clean (`djangocms_blog/cms_wizards.py`):

```python
registered_wizards = set()
for config in BlogConfig.objects.all().order_by('namespace'):
    new_wizard = type(str(slugify(config.app_title)), (PostWizard,), {})
    new_form = type(str('{0}Form').format(slugify(config.app_title)), (PostWizardForm,), {
        'default_appconfig': config.pk
    })
    post_wizard = new_wizard(
        title=_('New {0}').format(config.object_name),
        weight=200,
        form=new_form,
        model=Post,
        description=_('Create a new {0} in {1}').format(config.object_name, config.app_title),
    )

    if post_wizard.id not in registered_wizards:
        try:
            registered_wizards.add(post_wizard.id)
            wizard_pool.register(post_wizard)
        except AlreadyRegisteredException:  # pragma: no cover
            if settings.DEBUG:
                raise  
```

**Comment by yakky** [comment]
@sephii thanks for the great analysis! Actually the id of the classes should be different because they are two different classes (generated using ``type`` at the very beginning of the ``for`` cycle)
With the above workaround you can then use the wizard on both sites and have the correct apphookconfig provided to the post instance when you create them?
I'll add the workaround to the next release, anyway.
Thanks for sharing your findindgs

**Comment by sephii** [comment]
Ok thanks to your last comment I realized what's happening. I didn't realize `app_title` was supposed to be unique so I gave my two `BlogConfig` objects the same `app_title` attribute (I named them both `Blog`), so they got the same generated id and this led to the exception being raised. Maybe in the end I shouldn't even use two different `BlogConfig` objects, I did it that way because the migration process from a preview version created these.

I made a test with the workaround I proposed and it works but it then only lets you use the wizard for the first `BlogConfig` object, which might be a problem. So I think even if the workaround gets integrated there should at least be a warning mentioning the duplicate.
